### ISSUE TITLE
Panic in libnetwork during daemon start (panic: runtime error: index out of range [0] with length 0)

### ISSUE THREAD

**Issue by neersighted** [moby/moby#45057]
### Description

_From https://forums.docker.com/t/dockerd-crashing-in-libnetwork-on-launch/134710_

> I’ve recently started seeing a weird dockerd crash in libnetwork on start. I tried downgrading, but that didn’t help. I also tried uninstalling the Nvidia container tools. The only solution that fixed the crash was completely wiping away /var/lib/docker. The only container I am running on this daemon is Jellyfin. However, the problem returns on system reboot.

### Reproduce

```
$ sudo dockerd
INFO[2023-02-19T17:29:03.455357713-05:00] Starting up
INFO[2023-02-19T17:29:03.456995192-05:00] [core] [Channel #1] Channel created           module=grpc
INFO[2023-02-19T17:29:03.457036179-05:00] [core] [Channel #1] original dial target is: "unix:///run/containerd/containerd.sock"  module=grpc
INFO[2023-02-19T17:29:03.457075584-05:00] [core] [Channel #1] parsed dial target is: {Scheme:unix Authority: Endpoint:run/containerd/containerd.sock URL:{Scheme:unix Opaque: User: Host: Path:/run/containerd/containerd.sock RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}  module=grpc
INFO[2023-02-19T17:29:03.457099058-05:00] [core] [Channel #1] Channel authority set to "localhost"  module=grpc
INFO[2023-02-19T17:29:03.457237128-05:00] [core] [Channel #1] Resolver state updated: {
  "Addresses": [
    {
      "Addr": "/run/containerd/containerd.sock",
      "ServerName": "",
      "Attributes": {},
      "BalancerAttributes": null,
      "Type": 0,
      "Metadata": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} (resolver returned new addresses)  module=grpc
INFO[2023-02-19T17:29:03.457335134-05:00] [core] [Channel #1] Channel switches to new LB policy "pick_first"  module=grpc
INFO[2023-02-19T17:29:03.457397451-05:00] [core] [Channel #1 SubChannel #2] Subchannel created  module=grpc
INFO[2023-02-19T17:29:03.457475959-05:00] [core] [Channel #1 SubChannel #2] Subchannel Connectivity change to CONNECTING  module=grpc
INFO[2023-02-19T17:29:03.457505585-05:00] [core] [Channel #1 SubChannel #2] Subchannel picks a new address "/run/containerd/containerd.sock" to connect  module=grpc
INFO[2023-02-19T17:29:03.457674493-05:00] [core] [Channel #1] Channel Connectivity change to CONNECTING  module=grpc
INFO[2023-02-19T17:29:03.457919766-05:00] [core] [Channel #1 SubChannel #2] Subchannel Connectivity change to READY  module=grpc
INFO[2023-02-19T17:29:03.457944513-05:00] [core] [Channel #1] Channel Connectivity change to READY  module=grpc
INFO[2023-02-19T17:29:03.458479222-05:00] [core] [Channel #4] Channel created           module=grpc
INFO[2023-02-19T17:29:03.458491565-05:00] [core] [Channel #4] original dial target is: "unix:///run/containerd/containerd.sock"  module=grpc
INFO[2023-02-19T17:29:03.458502737-05:00] [core] [Channel #4] parsed dial target is: {Scheme:unix Authority: Endpoint:run/containerd/containerd.sock URL:{Scheme:unix Opaque: User: Host: Path:/run/containerd/containerd.sock RawPath: OmitHost:false ForceQuery:false RawQuery: Fragment: RawFragment:}}  module=grpc
INFO[2023-02-19T17:29:03.458510812-05:00] [core] [Channel #4] Channel authority set to "localhost"  module=grpc
INFO[2023-02-19T17:29:03.458532703-05:00] [core] [Channel #4] Resolver state updated: {
  "Addresses": [
    {
      "Addr": "/run/containerd/containerd.sock",
      "ServerName": "",
      "Attributes": {},
      "BalancerAttributes": null,
      "Type": 0,
      "Metadata": null
    }
  ],
  "ServiceConfig": null,
  "Attributes": null
} (resolver returned new addresses)  module=grpc
INFO[2023-02-19T17:29:03.458550045-05:00] [core] [Channel #4] Channel switches to new LB policy "pick_first"  module=grpc
INFO[2023-02-19T17:29:03.458564683-05:00] [core] [Channel #4 SubChannel #5] Subchannel created  module=grpc
INFO[2023-02-19T17:29:03.458578569-05:00] [core] [Channel #4 SubChannel #5] Subchannel Connectivity change to CONNECTING  module=grpc
INFO[2023-02-19T17:29:03.458598016-05:00] [core] [Channel #4 SubChannel #5] Subchannel picks a new address "/run/containerd/containerd.sock" to connect  module=grpc
INFO[2023-02-19T17:29:03.458637110-05:00] [core] [Channel #4] Channel Connectivity change to CONNECTING  module=grpc
INFO[2023-02-19T17:29:03.458767345-05:00] [core] [Channel #4 SubChannel #5] Subchannel Connectivity change to READY  module=grpc
INFO[2023-02-19T17:29:03.458800679-05:00] [core] [Channel #4] Channel Connectivity change to READY  module=grpc
INFO[2023-02-19T17:29:03.474131360-05:00] [graphdriver] using prior storage driver: overlay2
INFO[2023-02-19T17:29:03.479738078-05:00] Loading containers: start.
INFO[2023-02-19T17:29:03.672352245-05:00] failed to read ipv6 net.ipv6.conf.<bridge>.accept_ra  bridge=docker0 syspath=/proc/sys/net/ipv6/conf/docker0/accept_ra
panic: runtime error: index out of range [0] with length 0

goroutine 1 [running]:
github.com/docker/docker/libnetwork.(*controller).reservePools(0x5626885d0300?)
	/go/src/github.com/docker/docker/libnetwork/controller.go:862 +0xb65
github.com/docker/docker/libnetwork.New({0xc000628230, 0x8, 0xe})
	/go/src/github.com/docker/docker/libnetwork/controller.go:233 +0x6bc
github.com/docker/docker/daemon.(*Daemon).initNetworkController(0xc000b3c000, 0xc000ccd590)
	/go/src/github.com/docker/docker/daemon/daemon_unix.go:851 +0x4e
github.com/docker/docker/daemon.(*Daemon).restore(0xc000b3c000)
	/go/src/github.com/docker/docker/daemon/daemon.go:478 +0x51d
github.com/docker/docker/daemon.NewDaemon({0x5626885e4b38?, 0xc000ea6640}, 0xc0008f5b80, 0xc0005d2090)
	/go/src/github.com/docker/docker/daemon/daemon.go:1085 +0x2b79
main.(*DaemonCli).start(0xc000cb8420, 0xc000b39490)
	/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:200 +0x9f6
main.runDaemon(...)
	/go/src/github.com/docker/docker/cmd/dockerd/docker_unix.go:14
main.newDaemonCommand.func1(0xc000960c00?, {0x562689965730?, 0x0?, 0x0?})
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:38 +0x5e
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc000960c00, {0xc000052240, 0x0, 0x0})
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:916 +0x862
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000960c00)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:1044 +0x3bd
github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:968
main.main()
	/go/src/github.com/docker/docker/cmd/dockerd/docker.go:102 +0x15d
```

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:47:01 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:47:01 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: nvidia runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-60-generic
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 10
 Total Memory: 11.41GiB
 Name: carthage.lan
 ID: ZBCI:OVG2:MZ4Z:YKDJ:Z3T4:EG2G:NUWN:OQ2J:AIZW:62EC:IVJL:BZGJ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
 ```

**Comment by thaJeztah** [comment]
Looks to be happening here; https://github.com/moby/moby/blob/bc3805a0a0d3b5bd3f0e6c69f46ac08dd53377c7/libnetwork/controller.go#L860-L862

Which appears to making the assumption that `ipamV4Config` and `ipamV4Info` have the same size. Wondering if that's actually verified

**Comment by thaJeztah** [comment]
Is this the code restoring that? Looks like there's no guarantee in that; https://github.com/moby/moby/blob/bc3805a0a0d3b5bd3f0e6c69f46ac08dd53377c7/libnetwork/network.go#L679-L688

**Comment by corhere** [comment]
While there is no code which verifies `len(ipamV4Config) == len(ipamV4Info)` on unmarshal, there is _some_ guarantee that the slices are the same length...sometimes. https://github.com/moby/moby/blob/f825537119dd8b6dd099d351d8be194fc0eeb1cf/libnetwork/network.go#L1624
When a new network object is instantiated, its `ipamV4Info` is unset until `(*network).ipamAllocate()` is called.

`ipamV4Config` can be set on a network by either passing `NetworkOptionIpam()` or by using config inheritance (`NetworkOptionConfigFrom()` -> `(*network).applyConfigurationTo()`). Config-only networks can have `NetworkOptionIpam()` applied and the `ipamAllocate()` call is skipped over before the network is persisted to the store, but the `(*controller).reservePools()` loop skips over config-only networks so that can't be the culprit. And for all other networks, the network is only persisted to the store iff `ipamAllocate()` succeeds. `ipamAllocate()` is a no-op for the "special" network drivers `host` and `null`, but it is impossible to create user-defined networks with those drivers and there is no way to get the daemon to instantiate the built-in networks with `NetworkOptionConfigFrom()` or `NetworkOptionIpam()`. I don't see a way to create a non-config-only network using `(*controller).NewNetwork()` which is persisted to the store with `len(ipamV4Config) > 0 && len(ipamV4Info) == 0`.

A more promising lead is `(*network).ipamRelease()`. It sets `ipamV4Info = nil`. It is called by `(*network).delete()`, which persists the network immediately after: https://github.com/moby/moby/blob/f825537119dd8b6dd099d351d8be194fc0eeb1cf/libnetwork/network.go#L1063-L1066
If something was to go wrong with deleting the network resources resulting in an early return, the `c.deleteFromStore()` call would be skipped, leaving a network persisted in the store with `len(ipamV4Info) == 0` and possibly `len(ipamV4Config) > 0`. That network would have `inDelete == true` but those networks aren't skipped over by `(*controller).reservePools()`. The code comment preceding the `reservePools()` call in `libnetwork.New()` seems to indicate that this is by design.

That's enough theorizing for one day. The forum user can consistently reproduce the crash from scratch so I'd like to get repro steps from them before digging any further.
> The only solution that fixed the crash was **completely wiping away /var/lib/docker.** The only container I am running on this daemon is Jellyfin. However, **the problem returns** on system reboot.

(emphasis mine)

**Comment by aksiksi** [comment]
Thanks for creating this issue, @neersighted!

So I haven't be able to reproduce the crash after multiple reboots, but it definitely did happen at least once after a reboot. I will update if I see this again and share a link to a copy of `/var/lib/docker/network/files/local-kv.db`. 

Output of `docker info` for now:

```
$ docker info
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: nvidia runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-60-generic
 Operating System: Ubuntu 22.04.2 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 10
 Total Memory: 11.41GiB
 Name: carthage.lan
 ID: ZBCI:OVG2:MZ4Z:YKDJ:Z3T4:EG2G:NUWN:OQ2J:AIZW:62EC:IVJL:BZGJ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by neersighted** [comment]
Thanks for the additional info! Do you recall what version you upgraded from, and did you have any custom networks configured?

**Comment by rtaylor-rossvideo** [comment]
Hi!

We've been running into this exact same crash, and we can reproduce it every time (on our internal stuff, so unfortunately you guys won't be able to run it).

We can reproduce it on all docker versions starting from 20.10.23 and above. On 20.10.22, we can't reproduce it.

I'll try to give a light explanation of our world - feel free to ask for more details if needed.
We're running Ubuntu 20.04. We kick off from a docker compose file, which has 2 custom networks defined (both using the bridge driver), and 2 containers that run and use those networks. One of those containers is a base component of our system with a lot of jobs, but one of them is handling network configuration, so the user is able to go in and configure various network settings through it. It retrieves and sends network configuration through NetworkManager's DBus interface with the host system. To reproduce the crash, we have a system that is configured with a static IP address and we change it to something else (we've been keeping it static but picking a DHCP address the server recently had), then we immediately restart our world (not the entire system), which consists of stopping all docker containers, docker-compose down, and finally stopping the docker service. Usually after a network configuration change a restart will automatically run (without user intervention) and the crash doesn't seem to happen. It's only after a network configuration change where the user immediately performs a user restart themselves (so doesn't wait for the automatic one) that the crash will occur.

That's not the only way we've ran into the crash BTW, but I am lacking in details for those other circumstances and they were never reproducible. I believe in most? of those other circumstances network configurations were made recently before a crash occurred but I'm not 100% sure.

The workaround to remove the local-kv.db file works for us and leads me to ask:
Is it safe to nuke the /var/lib/docker/network/files/local-kv.db file before starting docker every time (or rather, maybe nuke that file and then restart the docker service is best)? Is there anything to watch out for with removing that file - any trouble we can run into? That's probably what we'll do for our next release if a fix is far out so I just want to confirm.

Here's docker version and info, the backtrace, and a local-kv.db file that was retrieved after the crash:

<details><summary>docker version</summary>
<p>

```
Client: Docker Engine - Community
 Version:           23.0.2
 API version:       1.42
 Go version:        go1.19.7
 Git commit:        569dd73
 Built:             Mon Mar 27 16:16:18 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.2
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.7
  Git commit:       219f21b
  Built:            Mon Mar 27 16:16:18 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.19
  GitCommit:        1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

</p>
</details> 


<details><summary>docker info</summary>
<p>

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.17.2
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 15
  Running: 15
  Paused: 0
  Stopped: 0
 Images: 38
 Server Version: 23.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
 Kernel Version: 5.4.0-144-generic
 Operating System: Ubuntu 20.04.6 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 12
 Total Memory: 15.5GiB
 Name: SG30d042e20c0a
 ID: ec8e34d1-f96a-4eb0-b074-70d74bf95716
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false

WARNING: No swap limit support
```

</p>
</details>

<details><summary>Backtrace</summary>
<p>

```
dockerd[7488]: time="2023-04-03T15:23:33.292788705Z" level=info msg="[core] [Channel #4] Channel switches to new LB policy \"pick_first\"" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.292813923Z" level=info msg="[core] [Channel #4 SubChannel #5] Subchannel created" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.292837496Z" level=info msg="[core] [Channel #4 SubChannel #5] Subchannel Connectivity change to CONNECTING" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.292862876Z" level=info msg="[core] [Channel #4 SubChannel #5] Subchannel picks a new address \"/run/containerd/containerd.sock\" to connect" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.292865455Z" level=info msg="[core] [Channel #4] Channel Connectivity change to CONNECTING" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.293004949Z" level=info msg="[core] [Channel #4 SubChannel #5] Subchannel Connectivity change to READY" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.293029304Z" level=info msg="[core] [Channel #4] Channel Connectivity change to READY" module=grpc
dockerd[7488]: time="2023-04-03T15:23:33.302297921Z" level=info msg="[graphdriver] using prior storage driver: overlay2"
dockerd[7488]: time="2023-04-03T15:23:33.324947758Z" level=info msg="Loading containers: start."
dockerd[7488]: panic: runtime error: index out of range [0] with length 0
dockerd[7488]: goroutine 1 [running]:
dockerd[7488]: github.com/docker/docker/libnetwork.(*controller).reservePools(0x555d08a42fe0?)
dockerd[7488]: #011/go/src/github.com/docker/docker/libnetwork/controller.go:862 +0xb65
dockerd[7488]: github.com/docker/docker/libnetwork.New({0xc000a02230, 0x8, 0xe})
dockerd[7488]: #011/go/src/github.com/docker/docker/libnetwork/controller.go:233 +0x6bc
dockerd[7488]: github.com/docker/docker/daemon.(*Daemon).initNetworkController(0xc00097e900, 0xc001255b30)
dockerd[7488]: #011/go/src/github.com/docker/docker/daemon/daemon_unix.go:851 +0x4e
dockerd[7488]: github.com/docker/docker/daemon.(*Daemon).restore(0xc00097e900)
dockerd[7488]: #011/go/src/github.com/docker/docker/daemon/daemon.go:478 +0x51d
dockerd[7488]: github.com/docker/docker/daemon.NewDaemon({0x555d08a57838?, 0xc0007c6080}, 0xc0001d3080, 0xc00012a780)
dockerd[7488]: #011/go/src/github.com/docker/docker/daemon/daemon.go:1085 +0x2b79
dockerd[7488]: main.(*DaemonCli).start(0xc00025f3e0, 0xc000a41ce0)
dockerd[7488]: #011/go/src/github.com/docker/docker/cmd/dockerd/daemon.go:200 +0x9f6
dockerd[7488]: main.runDaemon(...)
dockerd[7488]: #011/go/src/github.com/docker/docker/cmd/dockerd/docker_unix.go:14
dockerd[7488]: main.newDaemonCommand.func1(0xc000ef8000?, {0xc000e994d0?, 0x3?, 0x3?})
dockerd[7488]: #011/go/src/github.com/docker/docker/cmd/dockerd/docker.go:38 +0x5e
dockerd[7488]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).execute(0xc000ef8000, {0xc000052090, 0x3, 0x3})
dockerd[7488]: #011/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:916 +0x862
dockerd[7488]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0xc000ef8000)
dockerd[7488]: #011/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:1044 +0x3bd
dockerd[7488]: github.com/docker/docker/vendor/github.com/spf13/cobra.(*Command).Execute(...)
dockerd[7488]: #011/go/src/github.com/docker/docker/vendor/github.com/spf13/cobra/command.go:968
dockerd[7488]: main.main()
dockerd[7488]: #011/go/src/github.com/docker/docker/cmd/dockerd/docker.go:102 +0x15d
```

</p>
</details>

[local-kv.db - link expires April 17th](https://rossvideoltd-my.sharepoint.com/:u:/g/personal/rtaylor_rossvideo_com1/Ee06zbVq1hNMmRSghocuK_0BLP-oQGKT-iTKSTl8boL5Cg?e=J4yEk8)

I also have more local-kv.db files from other systems that experienced the same issue if that helps (although I imagine them to all look quite similar).

Happy to run any tests or gather more information if needed.

**Comment by thaJeztah** [comment]
> We can reproduce it on all docker versions starting from 20.10.23 and above. On 20.10.22, we can't reproduce it.

Only had a very quick look to see if something stood out, but in case others want to look if anything could relate to that;

- changes in the engine between those releases; https://github.com/moby/moby/compare/v20.10.22...v20.10.23
- all changes in libnetwork between those releases (not just the vendored code); https://github.com/moby/libnetwork/compare/dcdf8f176d1e13ad719e913e796fb698d846de98...05b93e0d3a95952f70c113b0bc5bdb538d7afdd7
- golang changes between; https://github.com/golang/go/compare/go1.18.9...go1.18.10

There also have been updates to docker compose since the 20.10.22 release (in case it's relevant to how networks are used / created).

**Comment by corhere** [comment]
I didn't see anything interesting in the engine diff. The only notable change in the libnetwork which stood out to me is https://github.com/moby/libnetwork/commit/6f36306d8b17ee7e960cf42eed499b768ef3d9a9 as it changes the order of operations when deleting a network.

I dumped the local-kv.db @rtaylor-rossvideo provided, and it contains pretty much what I predicted: a persisted network with `inDelete: true`, and an `ipamV4Config` but no `ipamV4Info`.

<details><summary>Details</summary>
<p>

```sh
for key in $(bbolt keys ~/Downloads/local-kv.db libnetwork); do
  printf "\n----- %s\n" "$key"
  bbolt get ~/Downloads/local-kv.db libnetwork "$key" | cut -b 9- | jq .
done
```

```
----- docker/network/v1.0/bridge/8a82c2ade602c1c3dfd9d29d7fb27dcc2108ba4fe73ca25a9d3362c7ed8da630/
{
  "AddressIPv4": "192.168.222.1/24",
  "BridgeIfaceCreator": 1,
  "BridgeName": "br-8a82c2ade602",
  "ContainerIfacePrefix": "",
  "DefaultBindingIP": "<nil>",
  "DefaultBridge": false,
  "DefaultGatewayIPv4": "<nil>",
  "DefaultGatewayIPv6": "<nil>",
  "EnableICC": true,
  "EnableIPMasquerade": true,
  "EnableIPv6": false,
  "HostIP": "<nil>",
  "ID": "8a82c2ade602c1c3dfd9d29d7fb27dcc2108ba4fe73ca25a9d3362c7ed8da630",
  "InhibitIPv4": false,
  "Internal": false,
  "Mtu": 0
}

----- docker/network/v1.0/bridge/9af31a41df6d0691df05a83e2ae0a79306b323e850ac2ecf380574c9fa28ba83/
{
  "AddressIPv4": "172.18.0.1/16",
  "BridgeIfaceCreator": 2,
  "BridgeName": "docker0",
  "ContainerIfacePrefix": "",
  "DefaultBindingIP": "0.0.0.0",
  "DefaultBridge": true,
  "DefaultGatewayIPv4": "<nil>",
  "DefaultGatewayIPv6": "<nil>",
  "EnableICC": true,
  "EnableIPMasquerade": true,
  "EnableIPv6": false,
  "HostIP": "<nil>",
  "ID": "9af31a41df6d0691df05a83e2ae0a79306b323e850ac2ecf380574c9fa28ba83",
  "InhibitIPv4": false,
  "Internal": false,
  "Mtu": 1500
}

----- docker/network/v1.0/endpoint_count/

----- docker/network/v1.0/endpoint_count/1fbe0635c71a176ec1c4e95d872600dd1cdb07a288f453a47b7ae90a7209f276/
{
  "Count": 0
}

----- docker/network/v1.0/endpoint_count/8a82c2ade602c1c3dfd9d29d7fb27dcc2108ba4fe73ca25a9d3362c7ed8da630/
{
  "Count": 0
}

----- docker/network/v1.0/endpoint_count/924e3b92d415c043c2a684d9d743dd77c5937cdc92e98ea556541e50caa5f59b/
{
  "Count": 0
}

----- docker/network/v1.0/endpoint_count/9af31a41df6d0691df05a83e2ae0a79306b323e850ac2ecf380574c9fa28ba83/
{
  "Count": 0
}

----- docker/network/v1.0/network/1fbe0635c71a176ec1c4e95d872600dd1cdb07a288f453a47b7ae90a7209f276/
{
  "addrSpace": "",
  "attachable": false,
  "configFrom": "",
  "configOnly": false,
  "created": "2023-04-03T15:17:38.510412558Z",
  "enableIPv6": false,
  "generic": {
    "com.docker.network.generic": {}
  },
  "id": "1fbe0635c71a176ec1c4e95d872600dd1cdb07a288f453a47b7ae90a7209f276",
  "inDelete": false,
  "ingress": false,
  "internal": false,
  "ipamOptions": null,
  "ipamType": "default",
  "labels": null,
  "loadBalancerIP": "",
  "loadBalancerMode": "NAT",
  "name": "none",
  "networkType": "null",
  "persist": true,
  "postIPv6": false,
  "scope": "local"
}

----- docker/network/v1.0/network/8a82c2ade602c1c3dfd9d29d7fb27dcc2108ba4fe73ca25a9d3362c7ed8da630/
{
  "addrSpace": "LocalDefault",
  "attachable": true,
  "configFrom": "",
  "configOnly": false,
  "created": "2023-04-03T15:19:05.750981547Z",
  "enableIPv6": false,
  "generic": {
    "com.docker.network.enable_ipv6": false,
    "com.docker.network.generic": {}
  },
  "id": "8a82c2ade602c1c3dfd9d29d7fb27dcc2108ba4fe73ca25a9d3362c7ed8da630",
  "inDelete": true,
  "ingress": false,
  "internal": false,
  "ipamOptions": null,
  "ipamType": "default",
  "ipamV4Config": "[{\"PreferredPool\":\"192.168.222.0/24\",\"SubPool\":\"\",\"Gateway\":\"\",\"AuxAddresses\":null}]",
  "labels": {
    "com.docker.compose.network": "servicebus",
    "com.docker.compose.project": "sdp",
    "com.docker.compose.version": "1.25.0"
  },
  "loadBalancerIP": "",
  "loadBalancerMode": "NAT",
  "name": "sdp_servicebus",
  "networkType": "bridge",
  "persist": true,
  "postIPv6": false,
  "scope": "local"
}

----- docker/network/v1.0/network/924e3b92d415c043c2a684d9d743dd77c5937cdc92e98ea556541e50caa5f59b/
{
  "addrSpace": "",
  "attachable": false,
  "configFrom": "",
  "configOnly": false,
  "created": "2023-04-03T15:17:38.519776865Z",
  "enableIPv6": false,
  "generic": {
    "com.docker.network.generic": {}
  },
  "id": "924e3b92d415c043c2a684d9d743dd77c5937cdc92e98ea556541e50caa5f59b",
  "inDelete": false,
  "ingress": false,
  "internal": false,
  "ipamOptions": null,
  "ipamType": "default",
  "labels": null,
  "loadBalancerIP": "",
  "loadBalancerMode": "NAT",
  "name": "host",
  "networkType": "host",
  "persist": true,
  "postIPv6": false,
  "scope": "local"
}

----- docker/network/v1.0/network/9af31a41df6d0691df05a83e2ae0a79306b323e850ac2ecf380574c9fa28ba83/
{
  "addrSpace": "LocalDefault",
  "attachable": false,
  "configFrom": "",
  "configOnly": false,
  "created": "2023-04-03T15:19:02.391703245Z",
  "enableIPv6": false,
  "generic": {
    "com.docker.network.enable_ipv6": false,
    "com.docker.network.generic": {
      "com.docker.network.bridge.default_bridge": "true",
      "com.docker.network.bridge.enable_icc": "true",
      "com.docker.network.bridge.enable_ip_masquerade": "true",
      "com.docker.network.bridge.host_binding_ipv4": "0.0.0.0",
      "com.docker.network.bridge.name": "docker0",
      "com.docker.network.driver.mtu": "1500"
    }
  },
  "id": "9af31a41df6d0691df05a83e2ae0a79306b323e850ac2ecf380574c9fa28ba83",
  "inDelete": false,
  "ingress": false,
  "internal": false,
  "ipamOptions": null,
  "ipamType": "default",
  "ipamV4Config": "[{\"PreferredPool\":\"172.18.0.0/16\",\"SubPool\":\"\",\"Gateway\":\"172.18.0.1\",\"AuxAddresses\":{}}]",
  "ipamV4Info": "[{\"IPAMData\":\"{\\\"AddressSpace\\\":\\\"LocalDefault\\\",\\\"AuxAddresses\\\":{},\\\"Gateway\\\":\\\"172.18.0.1/16\\\",\\\"Pool\\\":\\\"172.18.0.0/16\\\"}\",\"PoolID\":\"LocalDefault/172.18.0.0/16\"}]",
  "labels": null,
  "loadBalancerIP": "",
  "loadBalancerMode": "NAT",
  "name": "bridge",
  "networkType": "bridge",
  "persist": true,
  "postIPv6": false,
  "scope": "local"
}
```

</p>
</details> 

My hunch is that something is causing the `n.deleteNetwork()` call to fail, resulting in the `(*network).delete()` call returning early, before it deletes the network object from the datastore. That would be consistent with the evidence: https://github.com/moby/libnetwork/commit/6f36306d8b17ee7e960cf42eed499b768ef3d9a9 changed the order of operations from
1. mark network for deletion
2. `n.deleteNetwork()` or return error
3. `n.ipamRelease()`

to

1. mark network for deletion
2. `n.ipamRelease()`
3. `n.deleteNetwork()` or return error

Before that commit a failed `deleteNetwork` would mark the persisted network object in the datastore for deletion but otherwise leave it in a consistent state; not so much after.

Reverting https://github.com/moby/libnetwork/commit/6f36306d8b17ee7e960cf42eed499b768ef3d9a9 isn't an option as it would break Windows networking. Perhaps removing the `c.updateToStore(n)` call after `n.ipamRelease()` would be sufficient to prevent the network from being persisted in an inconsistent state if network deletion fails.

> To reproduce the crash, we have a system that is configured with a static IP address and we change it to something else (we've been keeping it static but picking a DHCP address the server recently had), then we immediately restart our world (not the entire system), which consists of stopping all docker containers, docker-compose down, and finally stopping the docker service. Usually after a network configuration change a world restart will automatically run (without user intervention) and the crash doesn't seem to happen. It's only after a network configuration change where the user immediately performs a world restart themselves (so doesn't wait for the automatic one) that the crash will occur.

Sounds like something is going wrong when libnetwork tries to do _something_ at the same time as NetworkManager is reconfiguring the physical network interface. @rtaylor-rossvideo thank you so much for offering to run more tests! I have one which would help immensely with both confirming my hypothesis and identifying any secondary issues with libnetwork:
- [Enable daemon debugging](https://docs.docker.com/config/daemon/logs/#enable-debugging)
- Reproduce the crash
- Share logs of `dockerd` _and_ NetworkManager, preferably interleaved, across the time period from the network configuration changing to the daemon restarting and crashing.
  - **I am most interested in the pre-restart daemon logs and what NetworkManager is doing concurrently**

**Comment by corhere** [comment]
Thanks @rtaylor-rossvideo! The log you have shared is very helpful, however I didn't notice until just now that the error I was hoping to find won't be logged to the daemon log. That error is actually returned in the `DELETE /networks/{}` API response. Would it be possible to capture any error responses from deleting networks and share them here?

**Comment by rtaylor-rossvideo** [comment]
> Thanks @rtaylor-rossvideo! The log you have shared is very helpful, however I didn't notice until just now that the error I was hoping to find won't be logged to the daemon log. That error is actually returned in the `DELETE /networks/{}` API response. Would it be possible to capture any error responses from deleting networks and share them here?

Sure, how should I go about doing that? I believe docker-compose down would be responsible for those requests? (If so, is it a matter of collecting the output of that?)

**Comment by corhere** [comment]
If it's not already logged by default, you could try passing the `--verbose` flag to `docker-compose`. Are you using docker-compose v1 (Python, deprecated) or v2 (Go, built into the `docker` CLI)?

**Comment by rtaylor-rossvideo** [comment]
> If it's not already logged by default, you could try passing the `--verbose` flag to `docker-compose`. Are you using docker-compose v1 (Python, deprecated) or v2 (Go, built into the `docker` CLI)?

```
docker-compose --version
docker-compose version 1.25.0, build unknown
```

The script that runs docker-compose is run by a separate user from a systemd service. I think I removed that from `syslog` and confused myself as to why nothing was being logged :smile:. I'll try `--verbose` and see if that helps.

**Comment by rtaylor-rossvideo** [comment]
OK @corhere here's `syslog` again which I left mostly intact. The docker-compose (with `--verbose`) output can be seen from the `ross-river` process. Hopefully this is what you were looking for.

https://pastebin.com/QBU9HQ2U

**Comment by thaJeztah** [comment]
Uploaded them to github (in case they expire before we managed to have a look)

[dockerd-logs.txt](https://github.com/moby/moby/files/11153820/dockerd-logs.txt)

**Comment by corhere** [comment]
Yes, that's as good as we can get without modifying docker-compose or dockerd. There's no sign of the "failed to remove network" type error I was hoping for, which disproves my hypothesis. But I think I know what's actually going on.

<details><summary>It looks like `ross-river` is doing everything twice, concurrently</summary>
<p>

```
Apr  4 21:09:10 SGb8cb299b3b5b ross-river[11679]: sdp_ndi-receiver_15
Apr  4 21:09:10 SGb8cb299b3b5b ross-river[11677]: sdp_ndi-receiver_15
Apr  4 21:09:10 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:10.332639174Z" level=debug msg="Calling DELETE /v1.41/containers/41bafe369e23056cedddd0d7670b2db2ab2c9eea1531d085baec6b4b8e7e35dc?force=1"
Apr  4 21:09:10 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:10.378699302Z" level=debug msg="Calling HEAD /_ping"
Apr  4 21:09:10 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:10.378711025Z" level=debug msg="Calling HEAD /_ping"
Apr  4 21:09:10 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:10.379077422Z" level=debug msg="Calling DELETE /v1.42/containers/sdp_ndi-receiver_15?force=1"
Apr  4 21:09:10 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:10.379081642Z" level=debug msg="Calling DELETE /v1.42/containers/sdp_ndi-receiver_15?force=1"
Apr  4 21:09:10 SGb8cb299b3b5b ross-river[11784]: Error response from daemon: No such container: sdp_ndi-receiver_15
Apr  4 21:09:10 SGb8cb299b3b5b ross-river[11783]: Error response from daemon: No such container: sdp_ndi-receiver_15
```

</p>
</details> 

<details><summary>In particular, it appears to be invoking `docker-compose down` twice</summary>
<p>

```
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.config.find: Using configuration files: /opt/ross/river/sdp/MD.yaml
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.config.find: Using configuration files: /opt/ross/river/sdp/MD.yaml
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_HOSTNAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_NODE_ID variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_DEBUG_MIDPLANE variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_CTL_IF variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_PTP_IF variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_HIDE_SLOT variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_HIDE_SLOT_PREFIX variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_SLOT_NAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MFC_NODE_NAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The MAX_SLOTS variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The TIMING variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The PRODUCT variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The PRODUCT_VERSION variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The PRODUCT_HASH variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.config.environment.__getitem__: The LICENSING_MODE variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_HOSTNAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_NODE_ID variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_DEBUG_MIDPLANE variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_CTL_IF variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_PTP_IF variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_HIDE_SLOT variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_HIDE_SLOT_PREFIX variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_SLOT_NAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MFC_NODE_NAME variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The MAX_SLOTS variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The TIMING variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The PRODUCT variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The PRODUCT_VERSION variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The PRODUCT_HASH variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.config.environment.__getitem__: The LICENSING_MODE variable is not set. Defaulting to a blank string.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: docker.utils.config.find_config_file: Trying paths: ['/root/.docker/config.json', '/root/.dockercfg']
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: docker.utils.config.find_config_file: No config file found
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: docker.utils.config.find_config_file: Trying paths: ['/root/.docker/config.json', '/root/.dockercfg']
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: docker.utils.config.find_config_file: No config file found
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.763226771Z" level=debug msg="Calling GET /v1.25/version"
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: docker.utils.config.find_config_file: Trying paths: ['/root/.docker/config.json', '/root/.dockercfg']
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: docker.utils.config.find_config_file: No config file found
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: docker.utils.config.find_config_file: Trying paths: ['/root/.docker/config.json', '/root/.dockercfg']
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: docker.utils.config.find_config_file: No config file found
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.765128775Z" level=debug msg="Calling GET /v1.25/version"
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/version HTTP/1.1" 200 822
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.command.get_client: docker-compose version 1.25.0, build unknown
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: docker-py version: 4.1.0
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: CPython version: 3.8.10
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.command.get_client: Docker base_url: http+docker://localhost
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.command.get_client: Docker version: Platform={'Name': 'Docker Engine - Community'}, Components=[{'Name': 'Engine', 'Version': '23.0.2', 'Details': {'ApiVersion': '1.42', 'Arch': 'amd64', 'BuildTime': '2023-03-27T16:16:18.000000000+00:00', 'Experimental': 'false', 'GitCommit': '219f21b', 'GoVersion': 'go1.19.7', 'KernelVersion': '5.4.0-146-generic', 'MinAPIVersion': '1.12', 'Os': 'linux'}}, {'Name': 'containerd', 'Version': '1.6.20', 'Details': {'GitCommit': '2806fc1057397dbaeefbea0e4e17bddfbd388f38'}}, {'Name': 'runc', 'Version': '1.1.5', 'Details': {'GitCommit': 'v1.1.5-0-gf19387a'}}, {'Name': 'docker-init', 'Version': '0.19.0', 'Details': {'GitCommit': 'de40ad0'}}], Version=23.0.2, ApiVersion=1.42, MinAPIVersion=1.12, GitCommit=219f21b, GoVersion=go1.19.7, Os=linux, Arch=amd64, KernelVersion=5.4.0-146-generic, BuildTime=2023-03-27T16:16:18.000000000+00:00
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.verbose_proxy.proxy_callable: docker inspect_network <- ('sdp_backend')
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.768268160Z" level=debug msg="Calling GET /v1.25/networks/sdp_backend"
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/networks/sdp_backend HTTP/1.1" 200 567
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: urllib3.connectionpool._make_request: http://localhost:None "GET /v1.25/version HTTP/1.1" 200 822
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.cli.command.get_client: docker-compose version 1.25.0, build unknown
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: docker-py version: 4.1.0
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: CPython version: 3.8.10
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.cli.command.get_client: Docker base_url: http+docker://localhost
```

</p>
</details> 

The logs are consistent with two instances of `docker-compose down` racing to tear down the stack.

<details><summary>Racing to delete the `sdp_backend` network:</summary>
<p>

The "Calling DELETE..." debug logs from `dockerd` mark the API request being received, and the `urllib3.connectionpool._make_request` debug logs from `docker-compose` (via `ross-river`) mark the API response being received. Note the HTTP status code.

It looks like (the `docker-compose` process whose logs are being relayed through) thread 13920 wins the race to delete the network. While the daemon is working on deleting the network, 13921 also tries to delete it and gets a 404 response. A short time later the network delete which 13920 requested completes, receiving an HTTP 204 response.

```
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.network.remove: Removing network sdp_backend
[...]
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.824875461Z" level=debug msg="Calling DELETE /v1.25/networks/sdp_backend"
[...]
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.network.remove: Removing network sdp_backend
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.cli.verbose_proxy.proxy_callable: docker remove_network <- ('sdp_backend')
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.825453374Z" level=debug msg="releasing IPv4 pools from network sdp_backend (2d058bd4ea0e91d7cca71b141d1274bee820e164ad03fd593f077c802f762f5b)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.825462722Z" level=debug msg="ReleaseAddress(LocalDefault/192.168.221.0/24, 192.168.221.1)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.825474162Z" level=debug msg="Released address PoolID:LocalDefault/192.168.221.0/24, Address:192.168.221.1 Sequence:App: ipam/default/data, ID: LocalDefault/192.168.221.0/24, DBIndex: 0x0, Bits: 256, Unselected: 253, Sequence: (0xc0000000, 1)->(0x0, 6)->(0x1, 1)->end Curr:4"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.825493950Z" level=debug msg="ReleasePool(LocalDefault/192.168.221.0/24)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.825727686Z" level=debug msg="cleanupServiceDiscovery for network:2d058bd4ea0e91d7cca71b141d1274bee820e164ad03fd593f077c802f762f5b"
[...]
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.826429783Z" level=debug msg="Calling DELETE /v1.25/networks/sdp_backend"
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: urllib3.connectionpool._make_request: http://localhost:None "DELETE /v1.25/networks/sdp_backend HTTP/1.1" 404 44
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.network.remove: Network sdp_backend not found.
[...]
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: urllib3.connectionpool._make_request: http://localhost:None "DELETE /v1.25/networks/sdp_backend HTTP/1.1" 204 0
```

</p>
</details> 

<details><summary>The logs around deleting the `sdp_servicebus` network might be the smoking gun.</summary>
<p>

```
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.cli.verbose_proxy.proxy_callable: docker remove_network <- ('sdp_servicebus')
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.827587105Z" level=debug msg="Calling DELETE /v1.25/networks/sdp_servicebus"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828014931Z" level=debug msg="releasing IPv4 pools from network sdp_servicebus (26d176d590ebe13e76c680a444b7007714c426aecefc616adde1cf8740c668a1)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828037841Z" level=debug msg="ReleaseAddress(LocalDefault/192.168.222.0/24, 192.168.222.1)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828054538Z" level=debug msg="Released address PoolID:LocalDefault/192.168.222.0/24, Address:192.168.222.1 Sequence:App: ipam/default/data, ID: LocalDefault/192.168.222.0/24, DBIndex: 0x0, Bits: 256, Unselected: 253, Sequence: (0xc0000000, 1)->(0x0, 6)->(0x1, 1)->end Curr:4"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828065925Z" level=debug msg="ReleasePool(LocalDefault/192.168.222.0/24)"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828303694Z" level=debug msg="cleanupServiceDiscovery for network:26d176d590ebe13e76c680a444b7007714c426aecefc616adde1cf8740c668a1"
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Interface backend.IPv4 no longer relevant for mDNS.
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Leaving mDNS multicast group on interface backend.IPv4 with address 192.168.221.1.
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Withdrawing address record for fe80::42:40ff:fe64:923 on backend.
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Withdrawing address record for 192.168.221.1 on backend.
Apr  4 21:09:27 SGb8cb299b3b5b NetworkManager[2349]: <info>  [1680642567.9123] device (backend): state change: activated -> unmanaged (reason 'unmanaged', sys-iface-state: 'removed')
Apr  4 21:09:27 SGb8cb299b3b5b dbus-daemon[864]: [system] Activating via systemd: service name='org.freedesktop.nm_dispatcher' unit='dbus-org.freedesktop.nm-dispatcher.service' requested by ':1.39' (uid=0 pid=2349 comm="/usr/sbin/NetworkManager --no-daemon " label="unconfined")
Apr  4 21:09:27 SGb8cb299b3b5b systemd[1]: Starting Network Manager Script Dispatcher Service...
Apr  4 21:09:27 SGb8cb299b3b5b dbus-daemon[864]: [system] Successfully activated service 'org.freedesktop.nm_dispatcher'
Apr  4 21:09:27 SGb8cb299b3b5b systemd[1]: Started Network Manager Script Dispatcher Service.
Apr  4 21:09:27 SGb8cb299b3b5b nm-dispatcher[13945]: run-parts: failed to stat component /etc/network/if-post-down.d/avahi-daemon: No such file or directory
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.942519584Z" level=debug msg="/usr/sbin/iptables, [--wait -t nat -C POSTROUTING -s 192.168.221.0/24 ! -o backend -j MASQUERADE]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.943509647Z" level=debug msg="/usr/sbin/iptables, [--wait -t nat -D POSTROUTING -s 192.168.221.0/24 ! -o backend -j MASQUERADE]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.944468245Z" level=debug msg="/usr/sbin/iptables, [--wait -t nat -C DOCKER -i backend -j RETURN]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.945163948Z" level=debug msg="/usr/sbin/iptables, [--wait -t nat -D DOCKER -i backend -j RETURN]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.946112679Z" level=debug msg="/usr/sbin/iptables, [--wait -t nat -C POSTROUTING -m addrtype --src-type LOCAL -o backend -j MASQUERADE]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.947161410Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -i backend -o backend -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.947955340Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -i backend -o backend -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.948929011Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -i backend ! -o backend -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.949722591Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -i backend ! -o backend -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.950536248Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -o backend -j DOCKER]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.951360594Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -o backend -j DOCKER]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.952166133Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -o backend -j DOCKER]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.952953845Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -o backend -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.953882464Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C FORWARD -o backend -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.955078373Z" level=debug msg="/usr/sbin/iptables, [--wait -D FORWARD -o backend -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.956003972Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C DOCKER-ISOLATION-STAGE-1 -i backend ! -o backend -j DOCKER-ISOLATION-STAGE-2]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.956700305Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -D DOCKER-ISOLATION-STAGE-1 -i backend ! -o backend -j DOCKER-ISOLATION-STAGE-2]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.957456556Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -C DOCKER-ISOLATION-STAGE-2 -o backend -j DROP]"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.958162252Z" level=debug msg="/usr/sbin/iptables, [--wait -t filter -D DOCKER-ISOLATION-STAGE-2 -o backend -j DROP]"
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Interface br-26d176d590eb.IPv6 no longer relevant for mDNS.
Apr  4 21:09:27 SGb8cb299b3b5b avahi-daemon[863]: Leaving mDNS multicast group on interface br-26d176d590eb.IPv6 with address fe80::42:48ff:fea8:a872.
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: urllib3.connectionpool._make_request: http://localhost:None "DELETE /v1.25/networks/sdp_backend HTTP/1.1" 204 0
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.verbose_proxy.proxy_callable: docker remove_network -> None
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.network.remove: Removing network sdp_servicebus
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.cli.verbose_proxy.proxy_callable: docker remove_network <- ('sdp_servicebus')
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.960859027Z" level=debug msg="Calling DELETE /v1.25/networks/sdp_servicebus"
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: urllib3.connectionpool._make_request: http://localhost:None "DELETE /v1.25/networks/sdp_servicebus HTTP/1.1" 404 47
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13920]: compose.network.remove: Network sdp_servicebus not found.
Apr  4 21:09:27 SGb8cb299b3b5b systemd[1]: Stopping Docker Application Container Engine...
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.987722804Z" level=info msg="Processing signal 'terminated'"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.987791480Z" level=debug msg="daemon configured with a 15 seconds minimum shutdown timeout"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.987837326Z" level=debug msg="start clean shutdown of all containers with a 15 seconds timeout..."
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991420521Z" level=debug msg="Unix socket /var/run/docker/libnetwork/55d71602060c.sock doesn't exist. cannot accept client connections"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991453471Z" level=info msg="[core] [Channel #1] Channel Connectivity change to SHUTDOWN" module=grpc
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991467013Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel Connectivity change to SHUTDOWN" module=grpc
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991476950Z" level=info msg="[core] [Channel #1 SubChannel #2] Subchannel deleted" module=grpc
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991483260Z" level=info msg="[core] [Channel #1] Channel deleted" module=grpc
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991491084Z" level=debug msg="Cleaning up old mountid : start."
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991538229Z" level=info msg="stopping event stream following graceful shutdown" error="<nil>" module=libcontainerd namespace=moby
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991764157Z" level=debug msg="Cleaning up old mountid : done."
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991881848Z" level=debug msg="Clean shutdown succeeded"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.991891628Z" level=info msg="Daemon shutdown complete"
Apr  4 21:09:28 SGb8cb299b3b5b avahi-daemon[863]: Interface br-26d176d590eb.IPv4 no longer relevant for mDNS.
Apr  4 21:09:28 SGb8cb299b3b5b avahi-daemon[863]: Leaving mDNS multicast group on interface br-26d176d590eb.IPv4 with address 192.168.222.1.
Apr  4 21:09:28 SGb8cb299b3b5b avahi-daemon[863]: Withdrawing address record for fe80::42:48ff:fea8:a872 on br-26d176d590eb.
Apr  4 21:09:28 SGb8cb299b3b5b avahi-daemon[863]: Withdrawing address record for 192.168.222.1 on br-26d176d590eb.
Apr  4 21:09:28 SGb8cb299b3b5b NetworkManager[2349]: <info>  [1680642568.0489] device (br-26d176d590eb): state change: activated -> unmanaged (reason 'unmanaged', sys-iface-state: 'removed')
Apr  4 21:09:28 SGb8cb299b3b5b nm-dispatcher[13970]: run-parts: failed to stat component /etc/network/if-post-down.d/avahi-daemon: No such file or directory
Apr  4 21:09:28 SGb8cb299b3b5b systemd[1]: docker.service: Succeeded.
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13921]: compose.cli.errors.exit_with_error: Couldn't connect to Docker daemon at http+docker://localhost - is it running?
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13921]: If it's at a non-standard location, specify the URL with the DOCKER_HOST environment variable.
Apr  4 21:09:28 SGb8cb299b3b5b systemd[1]: Stopped Docker Application Container Engine.
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13967]: Warning: Stopping docker.service, but it can still be activated by:
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13967]:   docker.socket
Apr  4 21:09:28 SGb8cb299b3b5b root: INFO - RIVER: Starting River...
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13986]: Warning: Stopping docker.service, but it can still be activated by:
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13986]:   docker.socket
```

</p>
</details> 

Unlike the `sdp_backend` deletion, there is no HTTP 204 response for deleting `sdp_servicebus`. There is, however, this log from the same thread ID as the race winner:
```
Apr  4 21:09:27 SGb8cb299b3b5b ross-river[13921]: compose.cli.verbose_proxy.proxy_callable: docker remove_network <- ('sdp_servicebus')
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.827587105Z" level=debug msg="Calling DELETE /v1.25/networks/sdp_servicebus"
Apr  4 21:09:27 SGb8cb299b3b5b dockerd[5029]: time="2023-04-04T21:09:27.828014931Z" level=debug msg="releasing IPv4 pools from network sdp_servicebus (26d176d590ebe13e76c680a444b7007714c426aecefc616adde1cf8740c668a1)"
[...]
Apr  4 21:09:28 SGb8cb299b3b5b ross-river[13921]: compose.cli.errors.exit_with_error: Couldn't connect to Docker daemon at http+docker://localhost - is it running?
```

What I think is happening is that there are two actors concurrently racing to do the equivalent of `docker-compose down; systemctl stop docker.service`, which is resulting in the daemon being shut down while one of the networks is still in the process of being deleted. The daemon is not shutting down as gracefully as intended—exiting before all in-flight API requests complete—killing the network delete operation before it has the chance to delete the network object from the datastore. If true, there are two bugs in the daemon:
1. libnetwork persists the network object to the datastore in an inconsistent state in the middle of deleting the network, leaving a persisted network which crashes the daemon on restore unless it is allowed to run to completion.
2. The daemon doesn't [wait for in-flight requests to be handled to completion](https://pkg.go.dev/net/http#Server.Shutdown) before exiting.

To work around these issues, stop racing multiple `docker-compose down` instances or at least ensure that all instances run to completion before shutting down the daemon.

**Comment by corhere** [comment]
I reproduced the panic locally on v23.0.3 in a DinD instance on my second attempt.
```console
~ # for n in 1 2; do sh -c "docker compose down; pkill -TERM dockerd" & done
```

I'll add a third issue to the list: `DELETE` on a resource which is in the process of being deleted should probably block until the deletion completes rather than prematurely return 404.

**Comment by rtaylor-rossvideo** [comment]
Thanks for the investigation @corhere!

Indeed, the only way we can actually reproduce the issue so far is with 2 (or more I guess) restarts of our world going at once. I didn't explain that well at all in my original message, sorry about that.

I never got the exact details from our testers about what they were doing when these crashes occurred (i.e. if they triggered the 2x restart behaviour), but I don't remember seeing anything in the logs that indicated that. I don't know where those logs have ran off to, but it's probably safe to assume that's what happened and I overlooked it in the logs.

We certainly should prevent multiple restarts happening at once - I'll try to figure out the best way to do that and report back. I imagine that will workaround it as you say since single restarts even after network configurations are made have no issues.

**Comment by rtaylor-rossvideo** [comment]
I haven't been able to reproduce the issue after preventing concurrent restarts, so that seems to do the trick (and was a much needed change on our end anyway!). It'll undergo some more heavy testing over the next couple of days but if you don't hear from me, well, no news is good news!

Thanks for looking into it. I see you've got a reproducer now, but if you'd like me to test any fixes just let me know since it's easy enough to reproduce.
